### ISSUE TITLE
How to run alltoall?

### ISSUE THREAD

**Issue by ykk199** [aliyun/SimAI#44]
When I replaced ALLREDUCE in microAllReduce.txt with ALLTOALL, the following error was reported. Are there any other places that need to be modified?Thanks！

id: embedding_layer , depen: -1 , wg_comp_time: 1
id:  , depen: -1 , wg_comp_time: 1
id:  , depen: -1 , wg_comp_time: 1
id:  , depen: -1 , wg_comp_time: 1
id:  , depen: -1 , wg_comp_time: 1
id:  , depen: -1 , wg_comp_time: 1
type: HYBRID_TRANSFORMER_FWD_IN_BCKWD ,num passes: 1 ,lines: 6 compute scale: 1 ,comm scale: 1
stat path: ./ncclFlowModel_ ,total rows: 1 ,stat row: 0
CSV path and filename: ./ncclFlowModel_detailed_36.csv
CSV path and filename: ./ncclFlowModel_EndToEnd.csv
simulator run
chunk size is: 16777216 , size is: 16777216 , layer_num is: 0 , node: 0
info: all-to-all forward pass collective issued for layer: embedding_layer, involved dimensions:  1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
assert failed. cond="false", msg="We assume at least one NIC is alive", +0.000568010s 35 file=/root/SimAI/astra-sim-alibabacloud/extern/network_backend/ns3-interface/simulation/src/point-to-point/model/rdma-hw.ccassert failed. cond="false", msg="We assume at least one NIC is alive", , line=251
+0.000568010sterminate called without an active exception
 32 file=/root/SimAI/astra-sim-alibabacloud/extern/network_backend/ns3-interface/simulation/src/point-to-point/model/rdma-hw.cc, line=Aborted (core dumped)

**Comment by Huoyuan100861** [comment]
> When I replaced ALLREDUCE in microAllReduce.txt with ALLTOALL, the following error was reported. Are there any other places that need to be modified?Thanks！
> 
> id: embedding_layer , depen: -1 , wg_comp_time: 1 id: , depen: -1 , wg_comp_time: 1 id: , depen: -1 , wg_comp_time: 1 id: , depen: -1 , wg_comp_time: 1 id: , depen: -1 , wg_comp_time: 1 id: , depen: -1 , wg_comp_time: 1 type: HYBRID_TRANSFORMER_FWD_IN_BCKWD ,num passes: 1 ,lines: 6 compute scale: 1 ,comm scale: 1 stat path: ./ncclFlowModel_ ,total rows: 1 ,stat row: 0 CSV path and filename: ./ncclFlowModel_detailed_36.csv CSV path and filename: ./ncclFlowModel_EndToEnd.csv simulator run chunk size is: 16777216 , size is: 16777216 , layer_num is: 0 , node: 0 info: all-to-all forward pass collective issued for layer: embedding_layer, involved dimensions: 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, assert failed. cond="false", msg="We assume at least one NIC is alive", +0.000568010s 35 file=/root/SimAI/astra-sim-alibabacloud/extern/network_backend/ns3-interface/simulation/src/point-to-point/model/rdma-hw.ccassert failed. cond="false", msg="We assume at least one NIC is alive", , line=251 +0.000568010sterminate called without an active exception 32 file=/root/SimAI/astra-sim-alibabacloud/extern/network_backend/ns3-interface/simulation/src/point-to-point/model/rdma-hw.cc, line=Aborted (core dumped)

Yes, the current alltoall implementation in SimCCL is still under internal review and testing. It's not publicly available yet. You can use SimAI-Analytical to run MoE models or try the native alltoall implementation in astra-sim for the time being.

**Comment by Huoyuan100861** [comment]
> They haven't implemented ALLTOALL at all. You can only test ALLTOALL in Analytical mode. By the way, it seems indispensable to add at least one another comm type(except alltoall) before ALLTOALL in forward process.

Before AllToAll, other communications were handled by the framework layer (Megatron), which led to varying communication patterns across different architectures.

**Comment by 32HD** [comment]
现在是否实现了alltoall呢？

**Comment by Huoyuan100861** [comment]
> 现在是否实现了alltoall呢？

The current alltoall implementation is in Pull Request status. See: https://github.com/aliyun/SimAI/pull/78/commits

**Comment by maoshunyu** [comment]
> > 现在是否实现了alltoall呢？
> 
> The current alltoall implementation is in Pull Request status. See: https://github.com/aliyun/SimAI/pull/78/commits


运行workload_analytical.txt时，如果AS_LOG_LEVEL!=0，有时会报错`double free or corruption (fasttop)
Aborted`
如果AS_LOG_LEVEL=0则不会出现该报错
这个现象线程数越多越容易出现

**Comment by Huoyuan100861** [comment]
> > > 现在是否实现了alltoall呢？
> > 
> > 
> > The current alltoall implementation is in Pull Request status. See: https://github.com/aliyun/SimAI/pull/78/commits
> 
> 运行workload_analytical.txt时，如果AS_LOG_LEVEL!=0，有时会报错`double free or corruption (fasttop) Aborted` 如果AS_LOG_LEVEL=0则不会出现该报错 这个现象线程数越多越容易出现

运行的SimAI-Analytical还是SimAI-NS3？我们会关注这个问题～

**Comment by maoshunyu** [comment]
> > > > 现在是否实现了alltoall呢？
> > > 
> > > 
> > > The current alltoall implementation is in Pull Request status. See: https://github.com/aliyun/SimAI/pull/78/commits
> > 
> > 运行workload_analytical.txt时，如果AS_LOG_LEVEL!=0，有时会报错`double free or corruption (fasttop) Aborted` 如果AS_LOG_LEVEL=0则不会出现该报错 这个现象线程数越多越容易出现
> 
> 运行的SimAI-Analytical还是SimAI-NS3？我们会关注这个问题～

是NS3，另外报错不止double free还有prev_size等等

**Comment by zhangkun3918** [comment]
1、MockNcclGroup组划分的时候，EP的组划分为什么需要依赖TP组的划分？
2、想使用ALLTOALL_EP，对应的workload应该是什么样？
